scanpy/utils.py

```diff
@@ -563,6 +569,183 @@ def unique_categories(categories):
     categories = np.array(natsorted(categories, key=lambda v: v.upper()))
     return categories
 
+# --------------------------------------------------------------------------------
+# Plotting data helpers
+# --------------------------------------------------------------------------------
+
+# TODO: implement diffxpy method, make singledispatch
+def rank_genes_groups_df(
+    adata: AnnData,
+    group: str,  # Can this be something other than a str?
+    *,
+    key: str = "rank_genes_groups",
+    pval_cutoff: Optional[float] = None,
+    logfc_min: Optional[float] = None,
+    logfc_max: Optional[float] = None,
+    gene_symbols: Optional[str] = None
+) -> pd.DataFrame:
+    """
+    Get `rank_genes_groups` results in the form of a :class:`pd.DataFrame`.
+
+    Params
+    ------
+    adata
+        Object to get results from.
+    group
+        Which group (key from :func:`scanpy.tl.rank_genes_groups` `groupby`) to
+        return results from.
+    key
+        Key differential expression groups were stored under.
+    pval_cutoff
+        Minimum adjusted pval to return.
+    logfc_min
+        Minumum logfc to return.
+    logfc_max
+        Maximum logfc to return.
+    gene_symbols
+        Column name in `.var` DataFrame that stores gene symbols. Specifying 
+        this will add that column to the returned dataframe.
+
+    Example
+    -------
+    >>> pbmc = sc.datasets.pbmc68k_reduced()
+    >>> sc.tl.rank_genes_groups(pbmc, groupby="louvain", use_raw=True, n_genes=pbmc.shape[1])
+    >>> dedf = sc.utils.rank_genes_groups_df(pbmc, group="0")
+    """
+    d = pd.DataFrame()
+    for k in ['scores', 'names', 'logfoldchanges', 'pvals', 'pvals_adj']:
+        d[k] = adata.uns["rank_genes_groups"][k][group]
+    if pval_cutoff is not None:
+        d = d[d["pvals_adj"] < pval_cutoff]
+    if logfc_min is not None:
+        d = d[d["logfoldchanges"] > logfc_min]
+    if logfc_max is not None:
+        d = d[d["logfoldchanges"] < logfc_max]
+    if gene_symbols is not None:
+        d = d.join(adata.var[gene_symbols], on="names")
+    return d
+
+
+# Would an array be faster?
+@doc_params(raw_layer_params=_docs.doc_raw_layers)
+def obs_values(
```

Comment:
Is there some duplicatation with https://github.com/theislab/anndata/blob/34f4eb63710628fbc15e7050e5efcac1d7806062/anndata/base.py#L1464?

I think we could have a public function in AnnData for this purpose.

Reply:
Your implementation in combination with `obs_values_df` is definitely better. But I really think it should go into AnnData, next to `.to_df()` which right now just gives the data matrix, but one could give it the options you have below.

Reply:
Definitely duplication, I hadn't realized we had that function in AnnData and was just going off how we assign colors for scatter plots. I agree it's better to have one function for this, and AnnData makes sense for where to put it.

Do all the current argument make sense for AnnData? For example, what about `gene_symbols`? If AnnData is meant to be agnostic to datatype, I could handle resolving gene_symbols in `obs_values_df`?

Also, these functions have different conventions for `layers`. Which one should we standardize on?

Reply:
Right, `gene_symbols` doesn't make sense for AnnData! Your suggestion makes sense.

Maybe Fidel didn't use the original https://github.com/theislab/scanpy/blob/9b522f54e0f839e1a0c9874ca658400bfe79a894/scanpy/plotting/_anndata.py#L311 in his functions? Unfortunate if I didn't notice. I don't have strong opinion on the convention for layers as long as it's there. We could also have a Slack discussion with Fidel, if you think there might be issues.

Reply:
Ah, I was looking at this function: https://github.com/theislab/scanpy/blob/9e4e5ee02e04cf618872d9b098e24f0542e8b227/scanpy/plotting/_tools/scatterplots.py#L651-L736

I don't think there's an issue per-say, I just think it'd be easier for me to follow/ debug the plotting functions if they were a little more standardized. We pretty frequently want a dataframe (or at least aligned arrays) of values per cell or gene, but this is done in a variety of ways.

Reply:
On layers, it looks like scanpy uses `layer=None` as the default and anndata does `layer='X'` (via `find . -name "*.py" -exec grep -n "layer=" {} +`).

I prefer the `scanpy` style, since if someone specifies `layer='X'` and they actually have a layer named `'X'` they probably want to use that.

Reply:
Oh, yes, I also prefer the Scanpy style. @Koncopd built most of the layers for AnnData, any reason to move away from 'X'? Any reason why you chose to do it that way?

Reply:
> I don't think there's an issue per-say, I just think it'd be easier for me to follow/ debug the plotting functions if they were a little more standardized. We pretty frequently want a dataframe (or at least aligned arrays) of values per cell or gene, but this is done in a variety of ways.

100% agreed.

Reply:
@ivirshup Could you please give an example where we have `layer='X'`, i don't see it.
@falexwolf not sure, maybe i had loom as a model, there is a similar thimg there if i remember correctly.

Reply:
Ah, i see. I'm not sure i wrote these.